File: Searching and Sorting Algorithms/Quick Sort/quickSort.cpp
#include <algorithm>
#include <iostream>
#include <vector>

using type = int;

void printArray(const std::vector<type>& array) {
  for (const auto& v : array)
    std::cout << v << " ";
  std::cout << "\n";
}

int partition(std::vector<type>& array, int low, int high) {
  int pivot = array[high];
  // std::cout << "pivot: " << pivot << "\n";

  int greater_element_index = low - 1;

  for (int i = low; i < high; i++) {
    if (array[i] <= pivot) {
      greater_element_index++;
      // std::cout << array[greater_element_index] << " <-> " << array[i] <<
      // "\n";
      std::swap(array[greater_element_index], array[i]);
    }
  }

  // std::cout << array[greater_element_index + 1] << " <---> " << array[high]
  // <<
  // "\n";
  std::swap(array[greater_element_index + 1], array[high]);

  return (greater_element_index + 1);
}

void quickSort(std::vector<type>& array, int low, int high) {
  if (low < high) {
    int pivot = partition(array, low, high);
    // printArray(array);
    quickSort(array, low, pivot - 1);
    quickSort(array, pivot + 1, high);
  }
}

int main() {
  std::vector<type> array = {8, 7, 6, 1, 0, 9, 2};

  std::cout << "Unsorted Array: \n";
  printArray(array);

  // perform quicksort on data
  quickSort(array, 0, array.size() - 1);

  std::cout << "Sorted array in ascending order: \n";
  printArray(array);
}

File: Searching and Sorting Algorithms/Quick Sort/quickSort.py

def printArray(array):
    for i in array:
        print(i, end=" ")
    print()


def partition(array, low, high):
    pivot = array[high]
    greater_element_index = low - 1
    for i in range(low, high):
        if array[i] <= pivot:
            greater_element_index += 1
            (array[greater_element_index], array[i]) = (
                array[i], array[greater_element_index])

    (array[greater_element_index+1], array[high]) = (
        array[high], array[greater_element_index+1])

    return greater_element_index+1


def quickSort(array, low, high):
    if low < high:
        pivot = partition(array, low, high)
        quickSort(array, low, pivot-1)
        quickSort(array, pivot+1, high)
    pass


if __name__ == '__main__':
    array = [8, 7, 6, 1, 0, 9, 2]
    print("Unsorted Array:\n", array)
    quickSort(array, 0, len(array)-1)
    print("Sorted array in ascending order:\n", array)
